topdown/http_test.go

```diff
@@ -2233,6 +2233,148 @@ func TestHTTPSNoClientCerts(t *testing.T) {
 	})
 }
 
+func TestCertSelectionLogic(t *testing.T) {
+	const (
+		localCaFile = "testdata/ca.pem"
+	)
+
+	// Set up Environment
+	caCertPEM, err := ioutil.ReadFile(localCaFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	caPool := x509.NewCertPool()
+	if ok := caPool.AppendCertsFromPEM(caCertPEM); !ok {
+		t.Fatal("failed to parse CA cert")
+	}
+
+	ca, err := ioutil.ReadFile(localCaFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = os.Setenv("CLIENT_CA_ENV", string(caCertPEM))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	getClientTLSConfig := func(obj ast.Object) *tls.Config {
+		_, client, err := createHTTPRequest(BuiltinContext{Context: context.Background()}, obj)
+		if err != nil {
+			t.Fatalf("Unexpected error creating HTTP request %v", err)
+		}
+		if client.Transport == nil {
+			return nil
+		}
+		return client.Transport.(*http.Transport).TLSClientConfig
+	}
+
+	systemCertsPool, err := x509.SystemCertPool()
+	if err != nil {
+		t.Fatalf("Unexpected error reading system certs %v", err)
+	}
+
+	tempSystemCertsPool, err := x509.SystemCertPool()
+	if err != nil {
+		t.Fatalf("Unexpected error reading system certs %v", err)
+	}
+	systemCertsAndCaPool, err := addCACertsFromBytes(tempSystemCertsPool, ca)
+	if err != nil {
+		t.Fatalf("Unexpected error merging system certs and ca certs %v", err)
+	}
+
+	tests := []struct {
+		note     string
+		input    map[*ast.Term]*ast.Term
+		expected [][]byte
+		msg      string
+		wantNil  bool
+	}{
+		{
+			note:     "tls_use_system_certs set to true",
+			input:    map[*ast.Term]*ast.Term{ast.StringTerm("tls_use_system_certs"): ast.BooleanTerm(true)},
+			expected: systemCertsPool.Subjects(),
+			msg:      "Expected TLS config to use system certs",
+			wantNil:  false,
+		},
+		{
+			note:     "tls_use_system_certs set to nil",
+			input:    map[*ast.Term]*ast.Term{ast.StringTerm("tls_use_system_certs"): ast.BooleanTerm(false)},
+			expected: nil,
+			msg:      "Expected no TLS config",
+			wantNil:  true,
```

Comment:
The only case where `wantNil` is true is the one where `expected` is `nil`. What do you think, we could get rid of `wantNil`, couldn't we? 🤔

Reply:
Right, `wantNil` is only true in one case. However, that's also the only case where the test-check is different. I thought about it a bit. If we decide to remove `wantNil`, then I can't include that particular test case in the loop anymore.

Reply:
What he’s saying is that you’re already testing expected == nil, so wantNil doesn’t really add anything.